### PR TITLE
Add TensorScatterUpdate, TensorScatterAdd and TensorScatterSub

### PR DESCRIPTION
**Pull request by PatriceVignola**

### CODE REVIEW

File: tensorflow/core/kernels/dml_tensor_scatter_ops.cc [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    result = BinaryOp()(input, result);
+
+    Microsoft::WRL::ComPtr<IDMLCompiledOperator> compiled_op =
+        scope.Compile(DML_EXECUTION_FLAG_NONE, {result});
+
+    Initialize(ctx, std::move(tensors), compiled_op.Get());
+  }
+
+  StatusOr<DmlGpuEvent> Compute(DmlKernelContext* ctx) const override {
+    D3D12BufferRegion input_buffer =
```
Comment (adtsai) [inline tensorflow/core/kernels/dml_tensor_scatter_ops.cc:298]: Can we eliminate all this custom logic by using FillValueConstant to generate the initial empty scalar?

Comment (PatriceVignola) [reply tensorflow/core/kernels/dml_tensor_scatter_ops.cc:298]: Sure we can. My thought process was that zeroing the buffer would be faster than adding another dispatch.

Comment (adtsai) [reply tensorflow/core/kernels/dml_tensor_scatter_ops.cc:298]: Maybe we should keep a pre-zero'd tensor attached to the device for situations such as these - I think we have several other instances where we do stuff like this. It's kinda similar to how CPUs have a zero register. Either way it's not a big deal, so I'll leave the decision up to you.

**Review by adtsai — APPROVED** [review]
:shipit:

File: tensorflow/core/kernels/dml_tensor_scatter_ops.cc [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+          .TypeConstraint<int64>("Tindices"),                               \
+      DmlKernelWrapper<DmlTensorScatterUpdateKernel<int64>,                 \
+                       GetOutputShapeAsInputShapeHelper>)                   \
+  REGISTER_KERNEL_BUILDER(                                                  \
+      Name("TensorScatterAdd")                                              \
+          .Device(DEVICE_DML)                                               \
+          .TypeConstraint<type>("T")                                        \
+          .TypeConstraint<int32>("Tindices"),                               \
+      DmlKernelWrapper<                                                     \
+          DmlTensorScatterBinaryKernel<int32, std::plus<dml::Expression>>,  \
```
Comment (jstoecker) [inline tensorflow/core/kernels/dml_tensor_scatter_ops.cc:355]: It's cool that this works. I know why it works (DMLX overloads), but it's still cool. :)

**Review by jstoecker — APPROVED** [review]

**Pull request merged**